include/alpaka/block/shared/st/BlockSharedMemStOmp5.hpp

```diff
@@ -44,12 +44,20 @@ namespace alpaka
             //-----------------------------------------------------------------------------
             static auto declareVar(BlockSharedMemStOmp5 const& smem) -> T&
             {
-#    pragma omp barrier
-#    pragma omp single
+                auto* data = blockSharedMemSt.template getVar<T>(TuniqueId);
+
+                if(!data)
                 {
-                    smem.alloc<T>();
+#    pragma omp barrier
+#    pragma onmp single
+                    {
+                        blockSharedMemSt.template alloc<T>(TuniqueId);
+                    }
+                    // lookup for the data chunk allocated by the master thread
+                    data = blockSharedMemSt.template getVar<T>(TuniqueId);
```

Comment:
Due to the `barrier` before the `single` no second allocation can happen before a thread reaches this line, so `getLatestVar`would be sufficient.

The barrier is placed at the start because it guards again crosstalk with the previous `declareVar`, but is actually only protects the calle to `getVar`/`getLatestVar` after the alloc, so it could also be placed there if you prefer.

Reply:
Same as my answer above. This is not possible because of the inlined dictionary inside the memory blob we use. Possible would be to return the memory from the function `blockSharedMemSt.template alloc<T>(TuniqueId);`

Reply:
This would not work because `alloc` is only called by the master.

Reply:
Yes, would only work for accelerators where the block extent is limited to one thread.